### PR TITLE
Capture exception with sentry when invariant conditions are met in migrations

### PR DESCRIPTION
**Pull request by danjm**
## Explanation

There are states that we expect to never see in migrations. Although we do handle such cases and return state unchanged when they occur. This PR will add sentry `captureExceptions` calls when these states are detected, to better inform debugging of bugs/errors where the presence of those states may play a factor.

## Manual Testing Steps

1. Build this branch, onboard
2. Run the following script in the background console
```
chrome.storage.local.get(({ data, meta }) => chrome.storage.local.set({ data: { ...data, PreferencesController: false }, meta: {...meta, version: 81} }, () => { window.location.reload() }))
```
3. You should see a network request to sentry with  `typeof state.PreferencesController is boolean` in the payload

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [ ] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: app/scripts/migrations/082.ts [hunk 1]
```diff
@@ -27,12 +27,69 @@ export async function migrate(originalVersionedData: {
...
-    !isObject(state.PreferencesController) ||
-    !isObject(state.NetworkController) ||
+    !isObject(state.PreferencesController)
+  ) {
+    global.sentry?.captureException?.(
+      new Error(
+        `typeof state.PreferencesController is ${typeof state.PreferencesController}`,
+      ),
+      {
+        // "extra" key is required by Sentry
```
Comment (Gudahtt) [inline app/scripts/migrations/082.ts:37]: The extra key is not required by Sentry

File: app/scripts/migrations/082.test.js [hunk 2]
```diff
@@ -1,6 +1,11 @@
 import { v4 } from 'uuid';
 import { migrate, version } from './082';
 
+const sentryCaptureExceptionMock = jest.fn();
+
+global.sentry = global.sentry || {};
```
Comment (Gudahtt) [inline app/scripts/migrations/082.test.js:6]: Is `global.sentry` ever set in this context?

Comment (danjm) [reply app/scripts/migrations/082.test.js:6]: oh good question...

no, I don't think it is

File: types/global.d.ts [hunk 3]
```diff
@@ -20,6 +20,8 @@ declare class SentryObject extends Sentry {
 
   // Calls either startSession or endSession based on optin status
   toggleSession: () => void;
+
+  captureException: (error: Error, extra?: Record<string, unknown>) => void;
```
Comment (Gudahtt) [inline types/global.d.ts:24]: Do we need to add this? It should already be part of the Sentry type being extended here.

File: app/scripts/migrations/082.test.js [hunk 4]
```diff
@@ -1,6 +1,11 @@
...
+global.sentry.captureException = sentryCaptureExceptionMock;
```
Comment (Gudahtt) [inline app/scripts/migrations/082.test.js:7]:
```suggestion
global.sentry = {
  captureException: sentryCaptureExceptionMock;
}
```

File: types/global.d.ts (hunk 3 above)
Comment (danjm) [reply types/global.d.ts:24]: If we remove it, we get errors like this:
```
app/scripts/migrations/082.ts:32:20 - error TS2339: Property 'captureException' does not exist on type 'SentryObject'.

32     global.sentry?.captureException?.(
                      ~~~~~~~~~~~~~~~~
```

File: app/scripts/migrations/082.test.js (hunk 2 above)
Comment (danjm) [reply app/scripts/migrations/082.test.js]: this redundancy has been removed in all places now

File: types/global.d.ts (hunk 3 above)
Comment (Gudahtt) [reply types/global.d.ts:24]: It looks like this is because this whole object is declared incorrectly. It's currently declared as a class, but it's not a class. It's an object. It should be defined like

```
type SentryObject = Sentry & {
  ...
}
```

Then we wouldn't need to re-add each method from Sentry itself that we're using.

File: app/scripts/migrations/082.ts [hunk 5]
```diff
@@ -27,12 +27,53 @@ export async function migrate(originalVersionedData: {
 function transformState(state: Record<string, unknown>) {
   if (
     !hasProperty(state, 'PreferencesController') ||
```
Comment (Gudahtt) [inline app/scripts/migrations/082.ts:29]: I've recently discovered that it is possible for the preferences controller to not be set in state. It doesn't get set to anything unless a state change is made. So it shouldn't be an error if it's missing - it just means that the user has never changed the preferences state.

Comment (Gudahtt) [reply app/scripts/migrations/082.ts:29]: See this snapshot for an example: https://github.com/MetaMask/metamask-extension/pull/20425

That snapshot is of the memstore not the persisted store. But the same should hold true of the persisted store since it's constructed in the same way (and the preferences controller is in both).

File: app/scripts/migrations/084.ts [hunk 6]
```diff
@@ -25,9 +25,26 @@ export async function migrate(originalVersionedData: {
...
+      new Error(
+        `typeof state.NetworkController is ${typeof state.NetworkController}`,
+      ),
+    );
+    return state;
+  }
+  if (!hasProperty(state.NetworkController, 'network')) {
+    const inPost077SupplementFor084state =
+      state.NetworkController.networkId !== undefined;
+    if (!inPost077SupplementFor084state) {
```
Comment (Gudahtt) [inline app/scripts/migrations/084.ts:40]: Nit: This was a little confusing due to the double negative. Could we flip this? i.e. check that it's undefined, rather than that it's "not not undefined"

**Review by Gudahtt — COMMENTED** [review]
Migrations 87, 88, 90, and 92 don't have any errors or console warnings in them. What was the reason for skipping those - was this PR intentionally scoped to just migrations related to the network controller?

**Comment by danjm** [comment]
> Migrations 87, 88, 90, and 92 don't have any errors or console warnings in them. What was the reason for skipping those - was this PR intentionally scoped to just migrations related to the network controller?

Not quite intentionally. I was scoping it to "controllers we know should definitely be defined"... but I guess that applies to the `TokensController` as well.

But does it apply to the `PhishingController`?

**Comment by Gudahtt** [comment]
I don't think any of the controllers in those unchanged migrations _necessarily_ will exist in state. Someone coming from a sufficiently old version would be missing them.

Though the validation for invalid _state properties_ would still apply. And for the cases where controller state was missing, I thought we had still planned to `log.warn` in those branches to say that we were skipping the migration due to missing state.

**Comment by danjm** [comment]
> Though the validation for invalid state properties would still apply. And for the cases where controller state was missing, I thought we had still planned to log.warn in those branches to say that we were skipping the migration due to missing state.

good points, I'll make the appropriate additions

**Comment by danjm** [comment]
> I don't think any of the controllers in those unchanged migrations necessarily will exist in state. Someone coming from a sufficiently old version would be missing them.

But the `TokensController` should have been added to state in migration 63, and the `TokenListController` should have been added to state in migration 77

**Comment by Gudahtt** [comment]
Yes that sounds right, I hadn't noticed those migrations worked that way. We don't typically do that sort of thing.

**Comment by Gudahtt** [comment]
Relates to #20449

File: types/global.d.ts (hunk 3 above)
Comment (danjm) [reply types/global.d.ts]: fixed in 494db5cb2f

File: app/scripts/migrations/082.ts (hunk 5 above)
Comment (danjm) [reply app/scripts/migrations/082.ts]: addressed in e692373382

File: app/scripts/migrations/084.ts (hunk 6 above)
Comment (danjm) [reply app/scripts/migrations/084.ts]: addressed in 1d4e534159

File: app/scripts/migrations/086.ts [hunk 7]
```diff
@@ -37,5 +37,24 @@ function transformState(state: Record<string, unknown>) {
...
   }
+  if (!isObject(state.NetworkController)) {
+    global.sentry?.captureException?.(
+      new Error(
+        `typeof state.NetworkController is ${typeof state.NetworkController}`,
+      ),
+    );
+  } else if (!hasProperty(state.NetworkController, 'provider')) {
+    const inPost077SupplementFor086state =
+      state.NetworkController.providerConfig !== undefined;
```
[the remark this answers is not in the document]
Comment (danjm) [reply app/scripts/migrations/086.ts]: addressed in 1d4e534159

**Comment by danjm** [comment]
@Gudahtt updates to 87,88,90 and 92 were added in ea7459f9b3

File: app/scripts/migrations/082.ts [hunk 8]
```diff
@@ -25,14 +25,55 @@ export async function migrate(originalVersionedData: {
 }
 
 function transformState(state: Record<string, unknown>) {
+  if (!hasProperty(state, 'PreferencesController')) {
+    return state;
```
Comment (Gudahtt) [inline app/scripts/migrations/082.ts:29]: Nit: could use a warning here

**Review by Gudahtt — DISMISSED** [review]
LGTM!

**Comment by Gudahtt** [comment]
Oh, lint error in one of the migrations

File: app/scripts/migrations/082.ts (hunk 8 above)
Comment (danjm) [reply app/scripts/migrations/082.ts:31]: done in a3aa9f2f69

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Review by mcmire — APPROVED** [review]
I didn't see messages from Sentry using the code posted in the PR description, but this worked for me. I also had to comment out the code that usually disables Sentry in dev mode. After that I saw a request to Sentry being made:

```
chrome.storage.local.get(({ data, meta }) => chrome.storage.local.set({ data: { ...data, NftController: false }, meta: {...meta, version: 81} }, () => { window.location.reload() }))
```

tl;dr: can confirm this works.

**Comment by danjm** [comment]
> I didn't see messages from Sentry using the code posted in the PR description, but this worked for me. I also had to comment out the code that usually disables Sentry in dev mode. After that I saw a request to Sentry being made

Yes, this makes sense because the PR was changed since I posted those manual steps to no longer throw an error in the case of preferences controller being undfefined.

Thanks @mcmire
